simulator/func_sim/multiplication.h

```diff
@@ -0,0 +1,126 @@
+    auto high_part = ( ( x1*y1) + (( x0*y1) >> halfwidth) + (( x1*y0) >> halfwidth) );
+    auto hi = narrow_cast<UT>( high_part + low_part);
+return hi;
+}
+
+template<typename T>
+auto riscv_multiplication_high_ss(T x, T y) {
+    using ST = sign_t<T>;
+    using UT = unsign_t<T>;
+    bool x_is_neg = x & ( UT{ 1} << (bitwidth<T> - 1));
+    bool y_is_neg = y & ( UT{ 1} << (bitwidth<T> - 1));
+    bool result_is_neg = x_is_neg ^ y_is_neg;
+    auto x_abs = ( x_is_neg) 
+                 ? ~( UT{ x} - 1)
+                 : UT{x};
+    auto y_abs = ( y_is_neg) 
+                 ? ~( UT{ y} - 1)
+                 : UT{y};
+    auto hi_abs = riscv_multiplication_high_uu( UT{ x_abs}, UT{ y_abs});
+    auto lo_abs = riscv_multiplication_low( UT{ x_abs}, UT{ y_abs});
+    auto result = UT{ hi_abs}; 
+    if( result_is_neg)
+        result = ( lo_abs == 0)
+                 ? narrow_cast<ST>( ~result + 1)
+                 : ~result;
+    else
+        result = narrow_cast<UT>( hi_abs);
+        
+    return result;
+}
+
+template<typename T>
+auto riscv_multiplication_high_su(T x, T y) {
+    using ST = sign_t<T>;
+    using UT = unsign_t<T>;
+    bool x_is_neg = x & ( UT{ 1} << (bitwidth<T> - 1));
+    auto x_abs = ( x_is_neg) 
+                 ? ~( UT{ x} - 1)
+                 : UT{x};
+    auto hi_abs = riscv_multiplication_high_uu( UT{ x_abs}, UT{ y});
+    auto lo_abs = riscv_multiplication_low( UT{ x_abs}, UT{ y});
+    auto result = UT{ hi_abs}; 
+    if( x_is_neg)
+        result = ( lo_abs == 0)
+                 ? narrow_cast<ST>( ~result + 1)
+                 : ~result;
+    else
+        result = narrow_cast<UT>( hi_abs);
+        
+    return result;
+}
+
+#else 
+// __SIZEOF_INT128__
+// using boost types
+template<typename T>
+auto riscv_multiplication_high_uu(T x, T y) {
+    using UT = unsign_t<T>;
+    using T2 = doubled_t<T>;
+    using UT2 = unsign_t<T2>;
+    auto value = narrow_cast<UT2>(T2{ x} * T2{ y});
+    auto hi = narrow_cast<UT>( value >> bitwidth<T>);
+    return hi;
+}
+
+template<typename T>
+auto riscv_multiplication_high_ss(T x, T y) { return riscv_multiplication_high_uu( x, y); }
+
+template<typename T>
+auto riscv_multiplication_high_su(T x, T y) { return riscv_multiplication_high_uu( x, y); }
+
+#endif
+
+template<typename T>
+auto riscv_division(T x, T y) {
+    using UT = unsign_t<T>;
+    if ( y == 0)
+        return narrow_cast<T>( all_ones<UT>());
+
+    if constexpr( !std::is_same_v<T, unsign_t<T>>)
+        if ( y == -1 && x == narrow_cast<T>(msb_set<unsign_t<T>>()))
+            return T{1} << ( bitwidth<T> - 1);
+
+    return x / y;
+}
+
+template<typename T>
+auto riscv_remainder(T x, T y) {
+    if ( y == 0)
+        return x;
+
+    if constexpr( !std::is_same_v<T, unsign_t<T>>)
+        if ( y == -1 && x == narrow_cast<T>(msb_set<unsign_t<T>>()))
+            return T{ 0};
+
+    return x % y;
+}
```

Comment:
Can MIPS implementations be merged here? Maybe we can share some code between the cores.

Reply:
I think that it is more logical to move ``riscv_div`` and ``riscv_rem`` to ``alu.h`` and substitute ``mips_div`` and ``mips_rem`` with combination of ``riscv_div`` and ``riscv_rem``

Reply:
ok

Reply:
Oh no, I was wrong. As I see, MIPS and RISC-V have different ISA in cases of division by zero and division overflow, so merging of ``riscv_div`` and ``mips_div`` will require additional conditional closes.

Reply:
ok, let's skip it.